src/Pool.sol

```diff
@@ -306,34 +408,197 @@ abstract contract Pool is Pausable, Ownable, ERC4626 {
+     * @param user_ Address of the user.
+     * @param depositId_ ID of the deposit.
+     */
+    function getUserDepositAmount(address user_, uint32 depositId_) external view onlyOwner returns (uint128) {
+        return _usersData[user_].deposits[depositId_].amount;
     }
 
     // *** Internal Functions ***
 
     /**
      * @dev Initializes the rewardsMultiplier array with the given parameters.
-     * @param blockNumbers Block numbers of the rewardsMultipliers.
-     * @param multipliers Multipliers of the rewardsMultipliers.
+     * @param blockNumbers_ Block numbers of the rewardsMultipliers.
+     * @param multipliers_ Multipliers of the rewardsMultipliers.
      */
-    function _initializeRewardsMultipliers(uint64[] memory blockNumbers, uint64[] memory multipliers) internal {
+    function _initializeRewardsMultipliers(uint64[] memory blockNumbers_, uint64[] memory multipliers_) internal {
         // Initialize the rewardsMultiplier array using the parameters.
-        for (uint256 i = 0; i < blockNumbers.length; i++) {
-            _rewardsMultipliers.push(RewardsMultiplier(blockNumbers[i], multipliers[i]));
+        for (uint256 i = 0; i < blockNumbers_.length; i++) {
+            _rewardsMultipliers.push(RewardsMultiplier(blockNumbers_[i], multipliers_[i]));
         }
     }
 
     /**
      * @dev Initializes the withdrawFees array with the given parameters.
-     * @param blockNumbers Block numbers of the withdrawFee.
-     * @param fees Fees of the withdrawFee.
+     * @param blockNumbers_ Block numbers of the withdrawFee.
+     * @param fees_ Fees of the withdrawFee.
      */
-    function _initializeWithdrawFees(uint64[] memory blockNumbers, uint64[] memory fees) internal {
+    function _initializeWithdrawFees(uint64[] memory blockNumbers_, uint64[] memory fees_) internal {
         // Initialize the withdrawFee array with the parameters.
-        for (uint256 i = 0; i < blockNumbers.length; i++) {
-            _withdrawFees.push(WithdrawFee(blockNumbers[i], fees[i]));
+        for (uint256 i = 0; i < blockNumbers_.length; i++) {
+            _withdrawFees.push(WithdrawFee(blockNumbers_[i], fees_[i]));
+        }
+    }
+
+    /**
+     * @dev Returns the withdraw fee corresponding to the blockNumber.
+     * @param blockNumber_ Block number of the deposit.
+     */
+    function _calcWithdrawFeePerBlock(uint64 blockNumber_) internal view returns (uint64) {
+        for (uint256 i = 0; i < _withdrawFees.length; i++) {
+            if (blockNumber_ < _withdrawFees[i].blockNumber) {
+                return _withdrawFees[i].fee;
+            }
         }
+        return 0;
+    }
+
+    /**
+     * @dev Updates the _rewardsPerToken variable.
+     */
+    function updateRewardsPerToken() internal {
+        uint128 currentBlock_ = uint128(block.number);
+
+        if (_totalAmount != 0) {
+            // Check current interval's rewards multiplier
+            for (uint256 i = 0; i < _rewardsMultipliers.length; i++) {
```

Comment:
info: do a while loop with ``while (currentBlock_ > rewardMultiplier_.blockNumber)` to go through all past blockIntervals and then a final piece of code to go through the current block interval.
gas: `_lastBlockUpdated``should be cached.